src/server/data_export_csv.py

```diff
@@ -0,0 +1,327 @@
+    payload.append([__('Event Leaderboards') + ': ' + __('Fastest 3 Consecutive Laps')])
+    for row in build_leaderboard(results['event_leaderboard'], primary_leaderboard='by_consecutives'):
+        payload.append(row)
+
+    payload.append([''])
+    payload.append([__('Class Leaderboards')])
+
+    # move unclassified heats to end
+    all_classes = sorted(list(results['heats_by_class'].keys()))
+    all_classes.append(all_classes.pop(all_classes.index(0)))
+
+    for class_id in all_classes:
+
+        valid_heats = False;
+        if len(results['heats_by_class'][class_id]):
+            for heat in results['heats_by_class'].keys():
+                if heat in results['heats']:
+                    valid_heats = True
+                    break
+
+        if valid_heats:
+            if class_id in results['classes']:
+                race_class = results['classes'][class_id]
+
+            payload.append([])
+            if race_class:
+                payload.append([__('Class') + ': ' + race_class['name']])
+                payload.append([])
+                payload.append([__('Class Summary')])
+                for row in build_leaderboard(race_class['leaderboard']):
+                    payload.append(row)
+            else:
+                if len(results['classes']):
+                    payload.append([__('Unclassified')])
+                else:
+                    payload.append([__('Heats')])
+
+            for heat_id in results['heats_by_class'][class_id]:
+                heat = results['heats'][heat_id]
+
+                payload.append([])
+                payload.append([__('Heat') + ': ' + heat['note']])
+
+                if len(heat['rounds']) > 1:
+                    payload.append([])
+                    payload.append([__('Heat Summary')])
+
+                    for row in build_leaderboard(heat['leaderboard']):
+                        payload.append(row)
+
+                for heat_round in heat['rounds']:
+                    payload.append([])
+                    payload.append([__('Round {0}').format(heat_round['id'])])
+
+                    for row in build_leaderboard(heat_round['leaderboard']):
+                        payload.append(row)
+
+                    payload.append([])
+                    payload.append([__('Round {0} Times').format(str(heat_round['id']))])
+
+                    for node in heat_round['nodes']:
+                        if len(node['laps']):
+                            laplist = []
+
+                            laplist.append(node['callsign'])
+
+                            for lap in node['laps']:
+                                if not lap['deleted']:
+                                    laplist.append(lap['lap_time_formatted'])
+
+                            payload.append(laplist)
+
+    return payload
+
+def discover(*args, **kwargs):
+    # returns array of exporters with default arguments
+    return [
+        {
+            'id': 'csv_pilots',
```

Comment:
This seems to be crying out for objects/classes. It seems args exists to facilitate this particular implementation. afaict a class with a name property and export function would provide sufficient contract.

Reply:
That's not a construct I'm familiar with. I'd be happy to learn more about this.